packages/xstate-wallet/src/workflows/tests/conclude-channel.test.ts

```diff
@@ -0,0 +1,218 @@
+import {interpret} from 'xstate';
+import waitForExpect from 'wait-for-expect';
+
+import {Init, machine as createChannel} from '../create-and-fund';
+import {machine as concludeChannel} from '../conclude-channel';
+
+import {Store} from '../../store';
+import {bigNumberify} from 'ethers/utils';
+
+import {
+  firstState,
+  calculateChannelId,
+  createSignatureEntry,
+  outcomesEqual
+} from '../../store/state-utils';
+import {ChannelConstants, Outcome, State} from '../../store/types';
+import {AddressZero} from 'ethers/constants';
+
+import {add} from '../../utils';
+
+import {
+  wallet1,
+  wallet2,
+  participants,
+  wallet3,
+  ledgerState,
+  first,
+  third,
+  second,
+  TEST_SITE,
+  budget
+} from './data';
+
+import {subscribeToMessages} from './message-service';
+
+import {FakeChain} from '../../chain';
+import {ETH_ASSET_HOLDER_ADDRESS, HUB} from '../../config';
+
+import {SimpleHub} from './simple-hub';
+import {TestStore} from './store';
+
+jest.setTimeout(20000);
+
+const EXPECT_TIMEOUT = process.env.CI ? 9500 : 2000;
+
+const chainId = '0x01';
+const challengeDuration = bigNumberify(10);
+const appDefinition = AddressZero;
+
+const targetChannel: ChannelConstants = {
+  channelNonce: bigNumberify(0),
+  chainId,
+  challengeDuration,
+  participants,
+  appDefinition
+};
+const targetChannelId = calculateChannelId(targetChannel);
+
+const destinations = participants.map(p => p.destination);
+
+const ledgerChannel: ChannelConstants = {
+  channelNonce: bigNumberify(1),
+  chainId,
+  challengeDuration,
+  participants,
+  appDefinition
+};
+
+const amounts = [bigNumberify(7), bigNumberify(5)];
+const ledgerAmounts = amounts.map(a => a.add(2));
+const depositAmount = ledgerAmounts.reduce(add).toHexString();
+
+const allocation: Outcome = {
+  type: 'SimpleAllocation',
+  assetHolderAddress: ETH_ASSET_HOLDER_ADDRESS,
+  allocationItems: [0, 1].map(i => ({
+    destination: destinations[i],
+    amount: amounts[i]
+  }))
+};
+
+const context: Init = {channelId: targetChannelId, funding: 'Direct'};
+
+let aStore: TestStore;
+let bStore: TestStore;
+
+const allSignState = (state: State) => ({
+  ...state,
+  signatures: [wallet1, wallet2].map(({privateKey}) => createSignatureEntry(state, privateKey))
+});
+
+let chain: FakeChain;
+
+const runUntilSuccess = async (machine, direct = true) => {
+  const runMachine = (store: Store) => interpret(machine(store).withContext(context)).start();
+  const [aService, bService] = [aStore, bStore].map(runMachine);
+
+  const targetState = direct ? 'success' : {virtualDefunding: 'asLeaf'};
+
+  await waitForExpect(async () => {
+    expect(bService.state.value).toEqual(targetState);
+    expect(aService.state.value).toEqual(targetState);
+  }, EXPECT_TIMEOUT);
+};
```

Comment:
An issue with this approach is that the `waitForExpect` can "miss" the target state. `waitForExpect` polls the callback until it passes, so if you transition past the target state during the polling interval, it might never pass.

You can do something like
```suggestion
  const services = [aStore, bStore].map(runMachine);
  const targetState = direct ? 'success' : {virtualDefunding: 'asLeaf'};

  await Promise.all(services.map(service => 
    new Promise(resolve => service.onTransition(state => state.matches(targetState) && resolve()))
  );
};
```

(I was not aware of this technique when I introduced `waitForExpect`)